File: app/src/main/java/com/staffaxis/hsm/domain/model/TarjaValores.kt
package com.staffaxis.hsm.domain.model

/**
 * Totales de un valor de tarja. El campo `minutesWorked` guarda un formato compuesto
 * ("H 8|C:33|Cajas 23 Cajones 55|AB:47573,53") y cada pantalla lo venia parseando por
 * su cuenta, con criterios distintos — de ahi salian las horas en 0 y la cosecha
 * contando empleados en vez de cachos. Este es el unico lugar donde se interpreta.
 */
data class TarjaValores(
    val horas: Float = 0f,
    val cosecha: Float = 0f,
    val cajas: Int = 0,
    val cajones: Int = 0,
    val importe: Float = 0f
) {
    operator fun plus(otro: TarjaValores) = TarjaValores(
        horas = horas + otro.horas,
        cosecha = cosecha + otro.cosecha,
        cajas = cajas + otro.cajas,
        cajones = cajones + otro.cajones,
        importe = importe + otro.importe
    )

    companion object {
        val CERO = TarjaValores()

        private val REGEX_CAJAS = Regex("Cajas ([0-9]+(?:[.,][0-9]+)?)")
        private val REGEX_CAJONES = Regex("Cajones ([0-9]+(?:[.,][0-9]+)?)")

        private fun num(s: String?): Float =
            s?.trim()?.replace(',', '.')?.toFloatOrNull() ?: 0f

        /**
         * Formatos que conviven en la base:
         *   "H 8"                    horas con prefijo (formato actual)
         *   "8" / "480"              numero plano (datos viejos: >16 se guardaba en minutos)
         *   "C:33" / "C"             cosecha — la vieja sin numero no aporta cachos
         *   "Cajas 23 Cajones 55"    cajas y cajones
         *   "AB:47573,53" / "$47573" importe
         */
        fun parse(minutesWorked: String?): TarjaValores {
            if (minutesWorked.isNullOrBlank()) return CERO
            var horas = 0f
            var cosecha = 0f
            var cajas = 0
            var cajones = 0
            var importe = 0f

            for (parte in minutesWorked.split("|").map { it.trim() }) {
                when {
                    parte.startsWith("H ") -> horas += num(parte.removePrefix("H "))
                    parte.startsWith("C:") -> cosecha += num(parte.removePrefix("C:"))
                    parte == "C" -> Unit // cosecha vieja, sin cantidad de cachos
                    parte.startsWith("AB:") -> importe += num(parte.removePrefix("AB:"))
                    parte.startsWith("$") -> importe += num(parte.removePrefix("$"))
                    parte.startsWith("Cajas") || parte.startsWith("Cajones") -> {
                        cajas += num(REGEX_CAJAS.find(parte)?.groupValues?.getOrNull(1)).toInt()
                        cajones += num(REGEX_CAJONES.find(parte)?.groupValues?.getOrNull(1)).toInt()
                    }
                    else -> {
                        val n = num(parte)
                        if (n > 0f) horas += if (n > 16f) n / 60f else n
                    }
                }
            }
            return TarjaValores(horas, cosecha, cajas, cajones, importe)
        }

        /** Suma los valores de varias tarjas. */
        fun sumar(valores: List<String?>): TarjaValores =
            valores.fold(CERO) { acc, mw -> acc + parse(mw) }
    }
}

File: web/src/domain/tarjaValores.ts
/**
 * Port exacto de domain/model/TarjaValores.kt.
 *
 * El campo `minutesWorked` guarda un formato compuesto
 * ("H 8|C:33|Cajas 23 Cajones 55|AB:47573,53") y este es el UNICO lugar donde se
 * interpreta, igual que en la app Kotlin.
 */

export interface TarjaValores {
  horas: number;
  cosecha: number;
  cajas: number;
  cajones: number;
  abonada: number;
}

export const CERO: TarjaValores = { horas: 0, cosecha: 0, cajas: 0, cajones: 0, abonada: 0 };

const REGEX_CAJAS = /Cajas ([0-9]+(?:[.,][0-9]+)?)/;
const REGEX_CAJONES = /Cajones ([0-9]+(?:[.,][0-9]+)?)/;

// Equivalente a Kotlin `s?.trim()?.replace(',', '.')?.toFloatOrNull() ?: 0f`.
// Ojo: toFloatOrNull es ESTRICTO (una cadena como "12abc" da null, no 12), asi que
// no alcanza con parseFloat — de ahi el chequeo con regex antes de convertir.
function num(s: string | null | undefined): number {
  if (s === null || s === undefined) return 0;
  const limpio = s.trim().replace(',', '.');
  if (!/^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$/.test(limpio)) return 0;
  const n = parseFloat(limpio);
  return Number.isNaN(n) ? 0 : n;
}

export function parse(minutesWorked: string | null | undefined): TarjaValores {
  if (minutesWorked === null || minutesWorked === undefined || minutesWorked.trim() === '') return CERO;

  let horas = 0;
  let cosecha = 0;
  let cajas = 0;
  let cajones = 0;
  let abonada = 0;

  for (const parte of minutesWorked.split('|').map((p) => p.trim())) {
    if (parte.startsWith('H ')) {
      horas += num(parte.slice(2));
    } else if (parte.startsWith('C:')) {
      cosecha += num(parte.slice(2));
    } else if (parte === 'C') {
      // cosecha vieja, sin cantidad de cachos
    } else if (parte.startsWith('AB:')) {
      abonada += num(parte.slice(3));
    } else if (parte.startsWith('$')) {
      // Formato viejo: la abonada sola, escrita con el signo adelante.
      abonada += num(parte.slice(1));
    } else if (parte.startsWith('Cajas') || parte.startsWith('Cajones')) {
      // Kotlin usa .toInt(), que TRUNCA hacia cero (no redondea).
      cajas += Math.trunc(num(REGEX_CAJAS.exec(parte)?.[1]));
      cajones += Math.trunc(num(REGEX_CAJONES.exec(parte)?.[1]));
    } else {
      const n = num(parte);
      // Datos viejos: si el numero plano es > 16 estaba guardado en minutos.
      if (n > 0) horas += n > 16 ? n / 60 : n;
    }
  }

  return { horas, cosecha, cajas, cajones, abonada };
}

export function sumarValores(a: TarjaValores, b: TarjaValores): TarjaValores {
  return {
    horas: a.horas + b.horas,
    cosecha: a.cosecha + b.cosecha,
    cajas: a.cajas + b.cajas,
    cajones: a.cajones + b.cajones,
    abonada: a.abonada + b.abonada,
  };
}

export function sumar(valores: Array<string | null | undefined>): TarjaValores {
  return valores.reduce<TarjaValores>((acc, mw) => sumarValores(acc, parse(mw)), CERO);
}

/**
 * Cosecha de un registro. Desde que se abrio en Cañadas / Raigon-Inv el dato vive
 * en sus columnas y el "C:" dejo de escribirse, asi que hay que mirar los dos
 * lados: el texto para las tarjas viejas y las columnas para las nuevas.
 *
 * No se duplica: una tarja vieja tiene el texto y los subtipos en null, y una
 * nueva tiene los subtipos y ningun "C:".
 */
export function cosechaDe(
  minutesWorked: string | null | undefined,
  t?: { cosechaCanadas: number | null; cosechaInv: number | null } | null,
): number {
  return parse(minutesWorked).cosecha + (t?.cosechaCanadas ?? 0) + (t?.cosechaInv ?? 0);
}
